File: src/c4/inter/Access.java
package inter;
import treewalker.TreeWalker;
import code.ArrayRefCode;
import code.DreiAdrCode;

import lexer.*;

/*
 * Access ist eine Unterklasse von Op und beschreibt einen Array-Zugriff. 
 * In den Instanzenvariablen array und index werden die beiden Operanden 
 * abgelegt. Die Instanzenvariablen enthalten den Ausdruck für das Array
 * und für den Index.
 */

public class Access extends Op {
	Expr array;
	Expr index;

	public Access(Expr a, Expr i) {
		super(new Word("[]", Tag.INDEX));
		array = a;
		index = i;
	}

	// Für TranformWalker
	public Access(Expr a, Expr i, Type p) {
		super(new Word("[]", Tag.INDEX));
		array = a;
		index = i;
		type = p;
	}

	public Expr getArray() {
		return array;
	}

	public Expr getIndex() {
		return index;
	}
	
	public void setIndex(Expr index) {
		this.index = index;
	}

	// für die Drei-Adress-Code Erzeugung
	public DreiAdrCode codeForValueTo(Id id) {
		return (new ArrayRefCode(id, this));
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkAccessNode(this, arg);
	}


}

File: src/c4/inter/Access.rs
use super::{
    DreiAdrCode, Expr, ExprData, ExpressionNode, Id, Node, NodeData, OpData, Tag, Token,
    TreeWalker, Type, Walk,
};

/// Array access expression (`array[index]`).
#[derive(Clone, Debug, PartialEq)]
pub struct Access {
    op: OpData,
    array: Expr,
    index: Expr,
}

impl Access {
    pub fn new(array: Expr, index: Expr) -> Self {
        Self {
            op: OpData::new(Token::new("[]", Tag::Index)),
            array,
            index,
        }
    }

    pub fn with_type(array: Expr, index: Expr, ty: Type) -> Self {
        let mut access = Self::new(array, index);
        access.set_type(ty);
        access
    }

    pub fn get_array(&self) -> &Expr {
        &self.array
    }
    pub fn get_index(&self) -> &Expr {
        &self.index
    }
    pub fn set_index(&mut self, index: Expr) {
        self.index = index;
    }

    pub fn code_for_value_to(&self, target: Id) -> DreiAdrCode {
        let array = self
            .array
            .as_id()
            .expect("array access must be lowered to an Id");
        let index = self
            .index
            .as_singleton()
            .expect("array index must be lowered to a Singleton");
        DreiAdrCode::ArrayRef {
            target,
            array,
            index,
        }
    }
}

impl Node for Access {
    fn node_data(&self) -> &NodeData {
        &self.op.expr.node
    }
}

impl ExpressionNode for Access {
    fn expr_data(&self) -> &ExprData {
        &self.op.expr
    }
    fn expr_data_mut(&mut self) -> &mut ExprData {
        &mut self.op.expr
    }
}

impl<P, R> Walk<P, R> for Access {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_access_node(self, arg)
    }
}
